src/volt/llvm/abi/sysvamd64.d

```diff
@@ -0,0 +1,522 @@
+			 }
+		 }
+	 }
+ 
+	 bool addElements(LLVMTypeRef[] theElements)
+	 {
+		 foreach (i, element; theElements) {
+			 LLVMTypeKind ekind = LLVMGetTypeKind(element);
+			 final switch(ekind) with (LLVMTypeKind) {
+			 case Void, Half, X86_FP80, FP128, PPC_FP128, Label,
+				 Function, Vector, Metadata, X86_MMX, Token:
+				 return false;
+			 case Array:
+				 auto len = LLVMGetArrayLength(element);
+				 auto base = LLVMGetElementType(element);
+				 auto types = new LLVMTypeRef[](len);
+				 foreach (ref type; types) {
+					 type = base;
+				 }
+				 if (!addElements(types)) {
+					 return false;
+				 }
+				 break;
+			 case Float:
+				 addSize(AMD64_SYSV_HALFWORD_SZ, /*floating*/true);
+				 break;
+			 case Double:
+				 addSize(AMD64_SYSV_WORD_SZ, /*floating*/true);
+				 break;
+			 case Integer:
+				 auto width = LLVMGetIntTypeWidth(element);
+				 addSize(width, /*floating*/false);
+				 break;
+			 case Pointer:
+				 addSize(AMD64_SYSV_WORD_SZ, /*floating*/false, LLVMGetElementType(element));
+				 break;
+			 case Struct:
+				 LLVMTypeRef[] types;
+				 classifyStructType(state, element, /*out*/types);
+				 if (types.length == 0) {
+					 return false;
+				 } else {
+					 addElements(types);
+				 }
+				 break;
+			 }
+		 }
+		 return true;
+	 }
+ 
+	 LLVMGetStructElementTypes(type, elements.ptr);
+	 if (!addElements(elements)) {
+		 return Classification.Memory;
+	 }
+ 
+	 if (alignment != 0) {
+		 sz = calcAlignment(alignment, sz);
+	 }
+ 
+	 if (sz > AMD64_SYSV_MAX_COERCIBLE_SZ) {
+		 return Classification.Memory;
+	 }
+ 
+	 LLVMTypeRef[] getSegmentType(size_t segment)
+	 {
+		 switch (floats[segment]) {
+		 case ONE_FLOAT:
+			 return [LLVMFloatTypeInContext(state.context)];
+		 case TWO_FLOATS:
+			 auto t = LLVMFloatTypeInContext(state.context);
+			 return [LLVMVectorType(t, 2)];
+		 case ONE_DOUBLE:
+			 return [LLVMDoubleTypeInContext(state.context)];
+		 default:
+			 if (pointees[segment] !is null) {
+				 return [LLVMPointerType(pointees[segment], 0)];
+			 }
+			 size_t segmentSize;
+			 if (segment == 0) {
+				 if (sz >= AMD64_SYSV_WORD_SZ) {
+					 segmentSize = AMD64_SYSV_WORD_SZ;
+				 } else {
+					 segmentSize = sz;
+				 }
+			 } else {
+				 segmentSize = sz - AMD64_SYSV_WORD_SZ;
+			 }
+			 return [LLVMIntTypeInContext(state.context, cast(uint)segmentSize)];
+		 }
+	 }
+ 
+	 if (sz <= AMD64_SYSV_WORD_SZ) {
+		 structTypes = getSegmentType(0);
+		 return Classification.CoercedStructSingle;
+	 }
+ 
+	 structTypes = getSegmentType(0) ~ getSegmentType(1);
+ 
+	 return Classification.CoercedStructDouble;
+ }
+ 
+ void sysvAmd64AbiCoerceArguments(State state, ir.CallableType ct, ref LLVMValueRef[] args)
+ {
+	 for (size_t i = 0; i < args.length; ++i) {
+		 // We need a pointer. If it is a value, alloca and store.
+		 if (ct.abiData[i].length != 0 &&
+			 LLVMGetTypeKind(LLVMTypeOf(args[i])) != LLVMTypeKind.Pointer) {
+			 auto _alloca = LLVMBuildAlloca(state.builder, LLVMTypeOf(args[i]), "");
```

Comment:
We might have a helper functions that builds Allocas at the top of the function, you don't want these in a inner loop.